src/Windows/Avalonia.Win32/WindowImpl.AppWndProc.cs

```diff
@@ -739,6 +743,28 @@ message switch
 
                             Imm32InputMethod.Current.CompositionChanged(currentComposition);
                         }
+                        
+                        if  (flags == (GCS.GCS_COMPATTR | GCS.GCS_COMPCLAUSE | GCS.GCS_COMPREADSTR | GCS.GCS_COMPSTR | GCS.GCS_CURSORPOS |
+                            GCS.GCS_DELTASTART | GCS.GCS_RESULTCLAUSE | GCS.GCS_RESULTREADCLAUSE | GCS.GCS_RESULTREADSTR |
+                            GCS.GCS_RESULTSTR))
+                        {
+                            var rs = Imm32InputMethod.Current.GetCompositionString(GCS.GCS_RESULTSTR);
+                            _ignoreWmChar = true;
+                            Imm32InputMethod.Current.IsComposing = false;
+                            if (!string.IsNullOrEmpty(rs) && ToInt32(wParam) >= 32)
+                            {
+                                e = new RawTextInputEventArgs(WindowsKeyboardDevice.Instance, timestamp, Owner, rs);
+                            }
+                                
+                            Imm32InputMethod.Current.Composition = null;
+                            if (Imm32InputMethod.Current.IsActive)
+                            {
+                                Imm32InputMethod.Current.Client.SetPreeditText(null);
+                            }
+                            Imm32InputMethod.Current.IsComposing = true;
+                            keybd_event((byte)VirtualKeyStates.VK_RIGHT, 0, KEYEVENTF_EXTENDEDKEY | 0, 0);
```

Comment:
Simulating keyboard events seems like a hack, especially since some IME methods don't use keyboard at all. 

Have you checked if other UI toolkits are doing the same?

Reply:
Yes, I understand... I also would like to avoid this keybd_event thing,,, but this tactics is often used to control IME when user require behavior or actions that are not available in the IME by design. I know various cases of application softwares for Windows used in Japan, that control IME using keybd_event (but sorry, I don't know on other UI toolkits... the best open source UI toolkit having mature Japanese text processing capability that I know is Avalonia),

Reply:
Could we detect a particular IME system and only apply the hack there? Right now the looks like something that could interfere with other IME providers